Fix and simplify overwrite mode

Freaky opened the pull request:
This fixes overwrite mode by replacing the line counting and iterated cursor repositioning with a clear screen and placing the cursor at 1,1 - exactly what GNU watch does.

I've also done some tidying, hoisting some iteration and filtering logic into dedicated generator functions which helps to further simplify the main loop.

Freaky committed:
Fix and simplify overwrite mode

Overwrite mode previously counted printed lines and then moved the
cursor up that many lines, which breaks if the number of datasets from
one iteration to the next goes down.  This made it basically useless in
nonzero mode.

Instead, simply copy what GNU watch does by clearing the screen and
moving the cursor to 1,1 each iteration.

Freaky committed:
Hoist dataset and diff iteration into generator functions

Add dataset_iter, diff_iter, and filter_diff_iter generator functions,
which together replace the respective logic that was in the main loop.

While I'm here reduce some unnecessary conditional grouping.

Freaky committed:
Use a dict comprehension for building the dataset list

We're already using these everywhere, and I quite like how the loop body
now basically starts by describing exactly what this generator yields.

Freaky committed:
Make dataset an optional argument

This adds the requirement to platform implementations to accept a false
pool argument signifying data for all pools should be gathered.

Update README with the new usage, and update the blurb to follow recent
changes.

Freaky committed:
Fix FreeBSD support in the multi-pool case

objsets are only pool unique, now we can gather information for more
than one pool at a time we must include that in the key we use for
gathering statistics.

Freaky committed:
Fix dataset filtering

Freaky committed:
Save a line in overwrite mode

jimsalterjrs commented:
hey @Freaky , looks like this needs a rebase? Seeing conflicts in both the script and the README.md. Thanks!

Freaky commented:
Rebased.

Freaky committed:
Pull the iterator delay into delay_iter()

This further simplifies the main loop by removing from it the
responsibility to sleep, allowing -y and -z modes to simply call
continue instead of having to have the entire output code indented.

jimsalterjrs commented:
Thanks!
